sdk/messaging/azservicebus/internal/amqpLinks.go

```diff
@@ -118,22 +118,42 @@ func (links *amqpLinks) ManagementPath() string {
 
 // recoverLink will recycle all associated links (mgmt, receiver, sender and session)
 // and recreate them using the link.linkCreator function.
-func (links *amqpLinks) recoverLink(ctx context.Context) error {
+func (links *amqpLinks) recoverLink(ctx context.Context, theirLinkRevision *uint64) error {
 	ctx, span := tab.StartSpan(ctx, SpanRecoverLink)
 	defer span.End()
 
 	links.mu.RLock()
 	closedPermanently := links.closedPermanently
+	ourLinkRevision := links.revision
 	links.mu.RUnlock()
 
 	if closedPermanently {
+		span.AddAttributes(tab.StringAttribute("outcome", "was_closed_permanently"))
 		return errClosedPermanently{}
 	}
 
+	if theirLinkRevision != nil && ourLinkRevision > *theirLinkRevision {
+		// we've already recovered past their failure.
+		span.AddAttributes(
+			tab.StringAttribute("outcome", "already_recovered"),
+			tab.StringAttribute("lock", "readlock"),
+			tab.StringAttribute("revisions", fmt.Sprintf("ours(%d), theirs(%d)", ourLinkRevision, *theirLinkRevision)),
+		)
+		return nil
+	}
+
 	links.mu.Lock()
 	defer links.mu.Unlock()
 
-	links.revision++
+	if theirLinkRevision != nil && ourLinkRevision > *theirLinkRevision {
```

Comment:
This duplicated bit, is this just double-checked locking pattern?

Reply:
Yeah, check once in the cheap 'read lock' and then check again in the expensive 'write lock'.